src/messages/LimitedQueue.hpp

```diff
@@ -2,296 +2,245 @@
+     */
+    std::vector<T> pushFront(const std::vector<T> &items)
+    {
+        std::unique_lock lock(this->mutex_);
 
-            this->chunks_ = newChunks;
+        size_t numToPush = std::min(items.size(), this->space());
+        std::vector<T> pushed;
+        pushed.reserve(numToPush);
 
-            if (this->chunks_->size() == 1)
-            {
-                this->lastChunkEnd_ += offset;
-            }
+        size_t f = items.size() - numToPush;
+        size_t b = items.size() - 1;
+        for (; f < items.size(); ++f, --b)
+        {
+            this->buffer_.push_front(items[b]);
+            pushed.push_back(items[f]);
         }
 
-        return acceptedItems;
+        return pushed;
     }
 
-    // replace an single item, return index if successful, -1 if unsuccessful
-    int replaceItem(const T &item, const T &replacement)
+    /**
+     * @brief Replace the needle with the given item
+     *
+     * @param[in] needle the item to search for
+     * @param[in] replacement the item to replace needle with
+     * @return the index of the replaced item, or -1 if no replacement took place
+     */
+    int replaceItem(const T &needle, const T &replacement)
     {
-        std::lock_guard<std::mutex> lock(this->mutex_);
-
-        int x = 0;
+        std::unique_lock lock(this->mutex_);
 
-        for (size_t i = 0; i < this->chunks_->size(); i++)
+        for (int i = 0; i < this->buffer_.size(); ++i)
         {
-            auto &chunk = this->chunks_->at(i);
-
-            size_t start = i == 0 ? this->firstChunkOffset_ : 0;
-            size_t end =
-                i == chunk->size() - 1 ? this->lastChunkEnd_ : chunk->size();
-
-            for (size_t j = start; j < end; j++)
+            if (this->buffer_[i] == needle)
             {
-                if (chunk->at(j) == item)
-                {
-                    auto newChunk = std::make_shared<Chunk>();
-                    newChunk->resize(chunk->size());
-
-                    for (size_t k = 0; k < chunk->size(); k++)
-                    {
-                        newChunk->at(k) = chunk->at(k);
-                    }
-
-                    newChunk->at(j) = replacement;
-                    this->chunks_->at(i) = newChunk;
-
-                    return x;
-                }
-                x++;
+                this->buffer_[i] = replacement;
+                return i;
             }
         }
-
         return -1;
     }
 
-    // replace an item at index, return true if worked
+    /**
+     * @brief Replace the item at index with the given item
+     *
+     * @param[in] index the index of the item to replace
+     * @param[in] replacement the item to put in place of the item at index
+     * @return true if a replacement took place
+     */
     bool replaceItem(size_t index, const T &replacement)
     {
-        std::lock_guard<std::mutex> lock(this->mutex_);
+        std::unique_lock lock(this->mutex_);
```

Comment:
Commenting on this instance, but referring to all uses of `unique_lock`: given that there's only one mutex to be locked & no early unlocking that occurs during the block, can't we use the simpler `lock_guard` (avoiding a little overhead from the extra features of `unique_lock`)?

Reply:
To me, it's mostly the semantics of `unique_lock` vs `shared_lock` that matter most - explicitly declaring whether or not we can share access to it